File: Misc/Sorts/JavaScript/countingSort.js
/*
Counting Sort

Counting sort is a sorting technique based on keys between a specific range.
It works by counting the number of objects having distinct key values
(kind of hashing). Then doing some arithmetic to calculate the position of
each object in the output sequence.
*/

function countingSort(arr) {
    var countArr = [];
    //: Count each element in the given array, and put count at appropriate index
    for (var idx = 0; idx < arr.length; idx++) {
        if (countArr[arr[idx]] > 0) {
            countArr[arr[idx]] += 1;
        } else {
            countArr[arr[idx]] = 1;
        }
    }
    //: Modify count array by adding up previous counts
    for (var idx = 1; idx < countArr.length; idx++) {
        if (countArr[idx] === undefined) {
            countArr[idx] = 0;
        }
        if (countArr[idx - 1] === undefined) {
            countArr[idx - 1] = 0;
        }
        countArr[idx] = countArr[idx] + countArr[idx - 1];
    }
    //: Iterate through given array and check against count array to see where it goes
    var sortedArr = [];
    for (var idx = arr.length - 1; idx >= 0; idx--) {
        countArr[arr[idx]]--;
        sortedArr[countArr[arr[idx]]] = arr[idx];
    }
    //: Return sorted array
    return sortedArr;
}

var arr1 = [1, 4, 1, 2, 7, 5, 2];
console.log(`Before sorting: \n ${arr1} \n`)
arr1 = countingSort(arr1);
console.log(`After sorting: \n ${arr1}`)

File: Sorts/Python/counting_sort.py
# Python 2.7
from utils import timer


def counting_sort(arr):
    #type: (list) -> list
    # Create count list
    max_val = max(arr)
    count_arr = [0 for _ in range(max_val+1)]
    output = [0 for _ in range(len(arr))]
    for val in arr:
        count_arr[val] += 1
    # print count_arr

    # Add up the counts
    for i in range(1, len(count_arr)):
        count_arr[i] += count_arr[i-1]
    # print count_arr

    # Sort based on the count arr
    for val in arr:
        pos = count_arr[val] - 1
        output[pos] = val
        count_arr[val] -= 1

    return output


ex1 = [2, 1, 4, 5, 7, 1, 3, 9, 9, 2]
ex2 = [9, 5, 6, 3, 2, 8, 1, 7, 10, 4]
ex3 = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
ex4 = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

for ts in (ex1, ex2, ex3, ex4):
    print "Before: %s" % ts

    with timer():
        ts = counting_sort(ts)

    print "After: %s\n" % ts
